rpc/src/lib.rs

```diff
@@ -206,15 +213,53 @@ impl<B, C, SC, P, CT, BE> EthApiT for EthApi<B, C, SC, P, CT, BE> where
 	}
 
 	fn balance(&self, address: H160, number: Option<BlockNumber>) -> Result<U256> {
+		let header = self.select_chain.best_chain()
+			.map_err(|_| internal_err("fetch header failed"))?;
+
+		let number_param: u32;
+
 		if let Some(number) = number {
-			if number != BlockNumber::Latest {
-				unimplemented!("fetch nonce for past blocks is not yet supported");
+			if let Some(block_number) = number.to_min_block_num() {
+				number_param = block_number.unique_saturated_into();
+			} else if number == BlockNumber::Latest {
+				number_param = header.number().clone().unique_saturated_into() as u32;
+			} else {
+				unimplemented!("only latest or block number are supported");
+			}
+		} else {
+			number_param = header.number().clone().unique_saturated_into() as u32;
+		}
+
+		let mut block_hash = None;
+		if let Ok(result) = self.client.header(BlockId::Number(number_param.into())) {
+			if let Some(header) = result {
+				block_hash = Some(header.hash());
+			}
+		}
+
+		if let Some(block_hash) = block_hash {
+			// StorageProvider prefix
+			let mut prefix = storage_prefix_build(b"EVM", b"Accounts");
+			// StorageMap blake2_128_concat key
+			let mut storage_key = blake2_128(address.as_bytes()).to_vec();
+			storage_key.extend_from_slice(address.as_bytes());
+			// Module and Storage prefix + StorageMap concat
+			prefix.extend(storage_key.clone());
+
+			let key = StorageKey(prefix);
+
+			if let Ok(Some(data)) = self.client.storage(
+				&BlockId::Hash(block_hash), 
+				&key) {
+				return Ok(
+					self.client
+						.runtime_api()
+						.account_decode(&BlockId::Hash(header.hash()), data.0)
```

Comment:
If we are reading the raw account data, what's the reason still calling into runtime?

Reply:
Not adding the pallet-evm as a dependency for accessing the EvmAccount

Reply:
Also as a side note, `Decode::decode` cannot infer the data type, in this case pallet-evm::Account needs to be declared as the result type.

Reply:
The round trip call to runtime API is the main consumption of performance. If we still call in runtime I'm not sure how this PR brings improvements, TBH.

Reply:
Indeed. As in the RPC module we already have the `frontier-rpc-primitives` dep, we could do a public re-export of the `pallet-evm::Account` from there, and avoid calling the Runtime doing that. Let me update the PR with that.